crates/pallet-messenger/src/lib.rs

```diff
@@ -0,0 +1,244 @@
+// Copyright (C) 2021 Subspace Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Pallet messenger used to communicate between domains and other blockchains.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#![forbid(unsafe_code)]
+#![warn(rust_2018_idioms, missing_debug_implementations)]
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+use codec::{Decode, Encode, MaxEncodedLen};
+pub use pallet::*;
+use scale_info::TypeInfo;
+use sp_core::U256;
+
+/// State of a channel.
+#[derive(Default, Debug, Encode, Decode, Clone, Eq, PartialEq, TypeInfo, MaxEncodedLen)]
+pub enum ChannelState {
+    /// Channel between domains is initiated but do not yet send or receive messages in this state.
+    #[default]
+    Initiated,
+    /// Channel is open and can send and receive messages.
+    Open,
+    /// Channel is closed and do not send or receive messages.
+    Closed,
+}
+
+/// Channel identity.
+pub type ChannelId = U256;
+
+/// Nonce used as an identifier and ordering of messages within a channel.
+/// Nonce is always increasing.
+pub type Nonce = U256;
+
+/// Channel describes a bridge to exchange messages between two domains.
+#[derive(Default, Debug, Encode, Decode, Clone, Eq, PartialEq, TypeInfo, MaxEncodedLen)]
+pub struct Channel {
+    /// Channel identifier.
+    pub(crate) channel_id: ChannelId,
+    /// State of the channel.
+    pub(crate) state: ChannelState,
+    /// Next inbox nonce.
+    pub(crate) next_inbox_nonce: Nonce,
+    /// Next outbox nonce.
+    pub(crate) next_outbox_nonce: Nonce,
+    /// Latest delivered outbox message nonce.
+    pub(crate) latest_delivered_message_nonce: Nonce,
+    /// Maximum outgoing non-delivered messages.
+    pub(crate) max_outgoing_messages: u64,
+}
+
+#[derive(Default, Debug, Encode, Decode, Clone, Eq, PartialEq, TypeInfo)]
+pub struct InitiateChannelParams {
+    max_outgoing_messages: u64,
+}
+
+#[frame_support::pallet]
+mod pallet {
+    use crate::{Channel, ChannelId, ChannelState, InitiateChannelParams, U256};
+    use frame_support::pallet_prelude::*;
+    use frame_system::pallet_prelude::*;
+    use sp_runtime::ArithmeticError;
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+        // Domain ID uniquely identifies a Domain
+        type DomainId: Parameter + Member + Default + Copy + MaxEncodedLen;
```

Comment:
Is DomainId necessary to be generic? I thought it'd be just u32 or a wrapper type of u32.

Reply:
I prefer not to define it here and there was not one to directly use from primtivies already. We can define it in a different PR though